src/bio/bio_config.c

```diff
@@ -0,0 +1,531 @@
+	if (tmp == NULL) {
+		D_ERROR("realloc error\n");
+		return -DER_NOMEM;
+	}
+
+	*list = tmp;
+	if (spdk_pci_addr_parse(*list + count, traddr) < 0) {
+		D_ERROR("Invalid address %s\n", traddr);
+		return -DER_INVAL;
+	}
+
+	opts->num_pci_addr++;
+	return 0;
+}
+
+static void *
+read_file(const char *filename, size_t *size)
+{
+	FILE *file = fopen(filename, "r");
+	void *data;
+
+	if (file == NULL) {
+		/* errno is set by fopen */
+		return NULL;
+	}
+
+	data = spdk_posix_file_load(file, size);
+	fclose(file);
+	return data;
+}
+
+static int
+read_config(const char *config_file, struct json_config_ctx *ctx)
+{
+	struct spdk_json_val *values = NULL;
+	void *json = NULL, *end;
+	ssize_t values_cnt, rc;
+	size_t json_size;
+
+	json = read_file(config_file, &json_size);
+	if (!json) {
+		D_ERROR("Read JSON configuration file %s failed, rc: %d\n",
+			config_file, errno);
+		return -DER_INVAL;
+	}
+
+	rc = spdk_json_parse(json, json_size, NULL, 0, &end,
+			     SPDK_JSON_PARSE_FLAG_ALLOW_COMMENTS);
+	if (rc < 0) {
+		D_ERROR("Parsing JSON configuration failed (%zd)\n", rc);
+		rc = -DER_INVAL;
+		goto err;
+	}
+
+	values_cnt = rc;
+	values = calloc(values_cnt, sizeof(struct spdk_json_val));
+	if (values == NULL) {
+		D_ERROR("Out of memory\n");
+		rc = -DER_NOMEM;
+		goto err;
+	}
+
+	rc = spdk_json_parse(json, json_size, values, values_cnt, &end,
+			     SPDK_JSON_PARSE_FLAG_ALLOW_COMMENTS);
+	if (rc != values_cnt) {
+		D_ERROR("Parsing JSON configuration failed (%zd)\n", rc);
+		rc = -DER_INVAL;
+		goto err;
+	}
+
+	ctx->json_data = json;
+	ctx->json_data_size = json_size;
+
+	ctx->values = values;
+	ctx->values_cnt = values_cnt;
+
+	return 0;
+err:
+	D_FREE(json);
+	D_FREE(values);
+	return rc;
+}
+
+static int
+load_vmd_subsystem_config(struct json_config_ctx *ctx, bool *vmd_enabled)
+{
+	struct config_entry	 cfg = {};
+
+	D_ASSERT(ctx->config_it != NULL);
+	D_ASSERT(vmd_enabled != NULL);
+
+	if (spdk_json_decode_object(ctx->config_it, jsonrpc_cmd_decoders,
+				    SPDK_COUNTOF(jsonrpc_cmd_decoders), &cfg)) {
+		D_ERROR("Failed to decode config entry\n");
+		return -DER_INVAL;
+	}
+
+	if (strcmp(cfg.method, "enable_vmd") == 0)
+		*vmd_enabled = true;
+
+	D_FREE(cfg.method);
+	return 0;
+}
+
+static int
+load_bdev_subsystem_config(struct json_config_ctx *ctx, bool vmd_enabled,
+			   struct spdk_env_opts *opts)
+{
+	struct config_entry	 cfg = {};
+	struct spdk_json_val	*key;
+	char			*traddr;
+	int			 rc = 0;
+
+	D_ASSERT(ctx->config_it != NULL);
+
+	if (spdk_json_decode_object(ctx->config_it, jsonrpc_cmd_decoders,
+				    SPDK_COUNTOF(jsonrpc_cmd_decoders), &cfg)) {
+		D_ERROR("Failed to decode config entry\n");
+		return -DER_INVAL;
+	}
+
+	if ((strcmp(cfg.method, "bdev_nvme_attach_controller") != 0) || (cfg.params == NULL))
+		goto out;
+
+	key = spdk_json_object_first(cfg.params);
+	while (key != NULL) {
+		if (spdk_json_strequal(key, "traddr")) {
+			traddr = spdk_json_strdup(json_value(key));
```

Comment:
Many traddrs are leaked, I think single temporary 'traddr' is sufficient and don't forget to free it.